File: Python/Linked-List-Cycle-II.py
#!/usr/bin/env python
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# @param head, a ListNode
# @return a list node
def detectCycle(self, head):
    if head == None:
        return None
    node1 = head
    node2 = head
    while node1 != None:
        node2 = node2.next
        node1 = node1.next
        if node1 == None:
            return None
        node1 = node1.next
        if node2 == node1:
            break
    if node1 == None:
        return None
    node2 = head
    while node2 != node1:
        node1 = node1.next
        node2 = node2.next
    return node2

File: C++/Linked-List-Cycle-II.cpp
    /**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow = head, *fast = head;
        while (fast != NULL && fast->next != NULL) {
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow) break;
        }
        if (fast == NULL || fast->next == NULL) return NULL;
        slow = head;
        while (slow != fast) {
            slow = slow->next;
            fast = fast->next;
        }
        return slow;
    }
};